[tiongMax] ip

tiongMax opened the pull request:
### Yolo the bot

Having trouble in memorising the tasks? Fret not, Yolo is hear to help!
With Yolo,  you do not have to remember by yourself. It is:

- ~~difficult~~ easy to use 👀
- user friendly 

> Why do it youself when someone else can do it for you? -myself

To use it:
1. download it from [here](https://github.com/tiongMax/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Key features:
- [ ] -  Store tasks
- [ ] -  Handle deadlines
- [ ] - Basically it can do what a to-do list can do

Here is where the `Duke` program starts:
```Java
public class Duke {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

tiongMax committed:
Attempt on level 0.

tiongMax committed:
Attempt on level 1.

tiongMax committed:
Attempt on level 2.

tiongMax committed:
Attempt on level 3.

tiongMax committed:
Attempt on level 3.

tiongMax committed:
Attempt on level 3.

tiongMax committed:
Attempt on level 4.

tiongMax committed:
Attempt on level 5.

tiongMax committed:
Attempt on level 6.

tiongMax committed:
Attempt to test the program by comparing input and output file.

tiongMax committed:
Revamp the previous code

tiongMax committed:
Add save functional on hard disk for task

tiongMax committed:
Merge branch 'branch-Level-7'

tiongMax committed:
Add date time feature for duke

tiongMax committed:
Merge branch 'branch-Level-8'

tiongMax committed:
Organise the code into a more OOP way

tiongMax committed:
Merge branch 'branch-A-MoreOOP'

tiongMax committed:
Organise the files into packages

tiongMax committed:
Merge branch 'branch-A-Packages'

tiongMax committed:
Merge branch 'add-gradle-support' of https://github.com/tiongMax/ip into A-Gradle

tiongMax committed:
Setup gradle

tiongMax committed:
Merge branch 'branch-A-Gradle'

tiongMax committed:
Add JUnit to test certain class and respective methods

tiongMax committed:
Merge branch 'branch-A-JUnit'

tiongMax committed:
Add javaDoc for all classes and methods

tiongMax committed:
Merge branch 'branch-A-JavaDoc'

tiongMax committed:
Merge branch 'master' into branch-A-CodingStandard

tiongMax committed:
Fix code according to the Java coding standard

tiongMax committed:
Merge branch 'branch-A-CodingStandard'

tiongMax committed:
Add search for keyword functionality

tiongMax committed:
Merge branch 'branch-Level-9'

yiwen101 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -17,18 +27,35 @@ public LocalDateTime getDueDate() {
+     * @return The format of the task in the text file.
+     */
     @Override
     public String contentLine() {
```
emm, seems that you forget to change this method into a verb

yiwen101 reviewed (commented):
Looks good.
Very comprehensive documentation
Reasonable and neat OOPs and project structure

tiongMax committed:
Setup for Level-10

tiongMax committed:
Add GUI for Duke bot.

tiongMax committed:
Merge branch 'branch-Level-10'

peiran18 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,336 @@
+            }
+        } else if (command.equals("delete")) {
+
+            if (!isValidDetail) {
+                throw new InvalidIndexException();
+            }
+            int index = Integer.parseInt(details);
+            if (index > 0 && index <= tasks.size()) {
+                result = delete(index, tasks, storage);
+            } else {
+                throw new InvalidIndexException();
+            }
+        } else if (command.equals("unmark")) {
+
+            if (!isValidDetail) {
+                throw new InvalidIndexException();
+            }
+            int index = Integer.parseInt(details);
+            if (index > 0 && index <= tasks.size()) {
+                result = unmark(index, tasks, storage);
+            } else {
+                throw new InvalidIndexException();
+            }
+        } else if (command.equals("todo")) {
+            result = addTodo(details, tasks, storage);
+        } else if (command.equals("deadline")) {
+            result = addDeadline(details, tasks, storage);
+        } else if (command.equals("event")) {
+            result = addEvent(details, tasks, storage);
+        } else if (command.equals("due")) {
+            result = checkTaskDue(details, tasks);
+        } else if (command.equals("find")) {
+            result = find(details, tasks);
+        } else if (command.equals("bye")) {
+            isExit();
+        } else {
+            throw new NoSuchCommandException();
+        }
+        return result;
+    }
```
I can understand the logical flow behind the block, but is there anyway we can make it more readable? Maybe extract out eg.mark to check for valid details and parsing instead of doing everything here?

ChangruHenryQian replied:
I know that sometimes it is inevitable to use large blocks of code. Would it be better to add some comment inside to increase readability?

peiran18 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,336 @@
+     * @param storage The storage object that contains the file for updating.
+     * @return The response of the bot after mark the task as done.
+     */
+    public String mark(int i, TaskList tasks, Storage storage) {
+
+        String result = "";
+        tasks.get(i - 1).mark();
+        result += Ui.showLine() + "\n" + " Nice! I've marked this task as done:" + "\n" + " " + tasks.get(i - 1).toString();
+        result += "\n";
+        result += Ui.showLine();
+        storage.writeInto(tasks);
+        return result;
+    }
```
I see you mark method here! Why don't you let this method call parse the details and settle the exceptions? (extension from prev comment)

peiran18 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,336 @@
+     * @param storage The storage object that contains the file for updating.
+     * @return The response of the bot when the task is deleted.
+     */
+    public String delete(int index, TaskList tasks, Storage storage) {
+
+        tasks.remove(index - 1);
+        storage.writeInto(tasks);
+        String result = Ui.showLine() + "\n" + "Noted. I've removed this task:" + "\n  " + tasks.get(index - 1);
+        result += "\nNow that you have " + tasks.size() + (tasks.size() < 2 ? " task" : " tasks") + " in the list.";
+        result += "\n";
+        result += Ui.showLine();
+        return result;
+    }
```
Do consider putting the commands into sub classes! eg.DeleteCommand (makes it easier for you to find and edit next time too)

peiran18 reviewed (commented):
Overall LGTM! The documentation makes it easy for me to understand what each method does :)

ChangruHenryQian commented on src/main/java/duke/exception/UnmatchedArgumentException.java:
```diff
@@ -0,0 +1,40 @@
+
+import duke.Ui;
+
+public class UnmatchedArgumentException extends Exception {
```
I like your implementation of multiple possible exceptions.

ChangruHenryQian commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,74 @@
+
+public class Duke {
+
+    /*
+        The task list of Duke.
+     */
+    private TaskList tasks;
+    /*
+        The interface for interaction with user.
+     */
+    private Ui ui;
+    /*
+        The file that is responsible for reading, writing and saving of the tasks.
+     */
+    private Storage storage;
+    /*
+        A parser that is used to parse the input of the user.
+     */
+    private Parser parser;
```
Great work on providing Java Doc even for private members.

ChangruHenryQian commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,118 @@
+package duke;
+
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.task.Todo;
+import duke.task.Event;
+import duke.task.Deadline;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
```
Would it be better to divide the import statement into different parts to avoid big chunks?

ChangruHenryQian reviewed (commented):
Good job!

tiongMax committed:
Refactor the command into different classes

tiongMax committed:
Merge branch 'branch-A-CodeQuality'

tiongMax committed:
Fix bugs for command and apply assertions

Fix the code wrong command with correct keyword included will lead to
unexpected outcome and add assertion statements to suitable spot.

Restructure and make changes to the guard clause.

tiongMax committed:
Merge pull request #1 from tiongMax/branch-A-Assertions

Fix bugs for command and apply assertions

tiongMax committed:
Improve code quality by removing redundant codes.

tiongMax committed:
Merge pull request #2 from tiongMax/branch-A-CodeQuality

Improve code quality by removing redundant codes and rename.

tiongMax committed:
Add sort command to the bot.

New command class Sort Command is added to handle the sort command to
sort the tasks in the task list alphabetically or reverse.
Comparator is used to assist the sorting of the tasks according to their
description.

tiongMax committed:
Merge pull request #3 from tiongMax/branch-BCD-Extension

Add sort command to the bot.

tiongMax committed:
Refactor the code by replacing some codes into a function

tiongMax committed:
Merge pull request #5 from tiongMax/branch-refactor-code

Refactor the code by replacing some codes into a function

tiongMax committed:
Remove the binary file

tiongMax committed:
Improve the Ui of the bot

tiongMax committed:
Improve the Ui of the bot ver.2

tiongMax committed:
Change readme into a user guide

Before using the application, user is able to read the user guide to
understand the basic operations to use the app.

GitHub flavored Markdown (GFMD) is used so that the readme looks more
engaging and interesting.

tiongMax committed:
Merge branch 'branch-A-UserGuide'

tiongMax committed:
Improve the method checkIndex

tiongMax committed:
Test

tiongMax committed:
Minor change

tiongMax committed:
Minor change

tiongMax committed:
Minor change

tiongMax committed:
Update the Ui screenshot from .PNG .png
